rust-symcrypt/src/rsa.rs

```diff
@@ -0,0 +1,643 @@
+    /// - `Ok(Vec<u8>)`: A vector of bytes containing the signature.
+    /// - `Err(String)`: An error message if the signing operation fails.
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, String>;
+}
+
+/// A trait for RSA signature verification.
+pub trait RsaVerify {
+    /// Verifies a message signature using the RSA public key.
+    ///
+    /// # Parameters
+    ///
+    /// - `message`: A byte slice representing the original message.
+    /// - `signature`: A byte slice representing the signature to be verified.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` which is:
+    /// - `Ok(bool)`: A boolean indicating whether the signature is valid (`true`) or not (`false`).
+    /// - `Err(String)`: An error message if the verification operation fails.
+    fn verify(&self, message: &[u8], signature: &[u8]) -> Result<bool, String>;
+}
+
+/// A trait for RSA encryption.
+pub trait RsaEncrypt {
+    /// Encrypts a plaintext message using the RSA public key.
+    ///
+    /// # Parameters
+    ///
+    /// - `plaintext`: A byte slice representing the plaintext message to be encrypted.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` which is:
+    /// - `Ok(Vec<u8>)`: A vector of bytes containing the ciphertext.
+    /// - `Err(String)`: An error message if the encryption operation fails.
+    fn encrypt(&self, plaintext: &[u8]) -> Result<Vec<u8>, String>;
+}
+
+/// A trait for RSA decryption.
+pub trait RsaDecrypt {
+    /// Decrypts a ciphertext message using the RSA private key.
+    ///
+    /// # Parameters
+    ///
+    /// - `ciphertext`: A byte slice representing the ciphertext message to be decrypted.
+    ///
+    /// # Returns
+    ///
+    /// A `Result` which is:
+    /// - `Ok(Vec<u8>)`: A vector of bytes containing the decrypted plaintext.
+    /// - `Err(String)`: An error message if the decryption operation fails.
+    fn decrypt(&self, ciphertext: &[u8]) -> Result<Vec<u8>, String>;
+}
+
+// Utility function to reduce common RSA allocation call
+fn allocate_rsa(n_primes: u32, n_bits_mod: u32) -> Result<symcrypt_sys::PSYMCRYPT_RSAKEY, SymCryptError> {
+    let rsa_params = symcrypt_sys::SYMCRYPT_RSA_PARAMS {
+        version: 1 as symcrypt_sys::UINT32, // No other version aside from version 1 is specified
+        nBitsOfModulus: n_bits_mod as symcrypt_sys::UINT32,
+        nPrimes: n_primes as symcrypt_sys::UINT32, // To generate a only a public key, 0 must be passed.
+        nPubExp: 1 as symcrypt_sys::UINT32,        // Value must be 1.
+    };
+    unsafe {
+        // SAFETY: FFI calls
+        let result = symcrypt_sys::SymCryptRsakeyAllocate(&rsa_params, 0);
+        if result == ptr::null_mut() {
+            return Err(SymCryptError::AuthenticationFailure)
+        }
+        Ok(result)
+    }
+}
+
+/// Utility function to store a `u64` into a new byte vector in big-endian format.
+fn store_msb_first_u64(value: u64, size_of_exp: u32) -> Result<Vec<u8>, SymCryptError> {
```

Comment:
>size_of_exp [](http://example.com/codeflow?start=35&length=11)

I think this should just be called `size` or something; this function can be used for other things than exponents